File: go_company_cache/api/includes/type/db/company_data/member_list.go
package company_data

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/getCompassUtils/go_base_frame/api/system/log"
	"github.com/getCompassUtils/go_base_frame/api/system/mysql"
	"strings"
)

const memberListTableKey = "member_list"

type MemberRow struct {
	UserId            int64  `json:"user_id"`
	Role              int32  `json:"role"`
	NpcType           int32  `json:"npc_type"`
	Permissions       int32  `json:"permissions"`
	CreatedAt         int32  `json:"created_at"`
	UpdatedAt         int32  `json:"updated_at"`
	CompanyJoinedAt   int32  `json:"company_joined_at"`
	LeftAt            int32  `json:"left_at"`
	FullNameUpdatedAt int32  `json:"fullname_updated_at"`
	FullName          string `json:"fullname"`
	MbtiType          string `json:"mbti_type"`
	ShortDescription  string `json:"short_description"`
	AvatarFileKey     string `json:"avatar_file_key"`
	Comment           string `json:"comment"`
	Extra             string `json:"extra"`
}

// GetMemberRow получаем пользователя по id
func (dbConn *DbConn) GetMemberRow(ctx context.Context, userId int64) (*MemberRow, error) {

	var row MemberRow

	// проверяем, что у нас имеется подключение к необходимой базе данных
	query := fmt.Sprintf("SELECT user_id, role, npc_type, permissions, created_at, updated_at, company_joined_at, left_at, full_name_updated_at, full_name, mbti_type, short_description, avatar_file_key, comment, extra FROM `%s` WHERE `user_id` = ? LIMIT ?", memberListTableKey)

	queryCtx, cancel := context.WithTimeout(ctx, mysql.QueryTimeout)
	defer cancel()

	err := dbConn.Conn.QueryRowContext(queryCtx, query, userId, 1).Scan(
		&row.UserId,
		&row.Role,
		&row.NpcType,
		&row.Permissions,
		&row.CreatedAt,
		&row.UpdatedAt,
		&row.CompanyJoinedAt,
		&row.LeftAt,
		&row.FullNameUpdatedAt,
		&row.FullName,
		&row.MbtiType,
		&row.ShortDescription,
		&row.AvatarFileKey,
		&row.Comment,
		&row.Extra,
	)

	if err != nil {

		if err == sql.ErrNoRows {

			log.Infof("Row with user_id %d not found", userId)
			return nil, nil
		}
		return nil, err
	}

	return &row, nil
}

// GetMemberList получаем пользователя по id
func (dbConn *DbConn) GetMemberList(ctx context.Context, userIdList []int64) ([]*MemberRow, error) {

	var plugForIn []string
	var queryArgs []interface{}
	for _, v := range userIdList {

		plugForIn = append(plugForIn, "?")
		queryArgs = append(queryArgs, v)
	}

	queryArgs = append(queryArgs, len(plugForIn))
	query := fmt.Sprintf("SELECT user_id, role, npc_type, permissions, created_at, updated_at, company_joined_at, left_at, full_name_updated_at, full_name, mbti_type, short_description, avatar_file_key, comment, extra FROM `%s` WHERE `user_id` IN (%s) LIMIT ?", memberListTableKey, strings.Join(plugForIn, ","))

	queryCtx, cancel := context.WithTimeout(ctx, mysql.QueryTimeout)

	results, err := dbConn.Conn.QueryContext(queryCtx, query, queryArgs...)
	defer func() { _ = results.Close(); cancel() }()

	if err != nil {

		log.Errorf(err.Error())
		return nil, err
	}

	memberList := make([]*MemberRow, 0, len(plugForIn))

	for results.Next() {

		var row MemberRow

		err = results.Scan(&row.UserId,
			&row.Role,
			&row.NpcType,
			&row.Permissions,
			&row.CreatedAt,
			&row.UpdatedAt,
			&row.CompanyJoinedAt,
			&row.LeftAt,
			&row.FullNameUpdatedAt,
			&row.FullName,
			&row.MbtiType,
			&row.ShortDescription,
			&row.AvatarFileKey,
			&row.Comment,
			&row.Extra,
		)
		if err != nil {
			return nil, err
		}

		memberList = append(memberList, &row)
	}
	return memberList, nil
}

File: php_monolith/src/Compass/Conversation/api/includes/gateway/db/company_data/member_list.php
<?php

namespace Compass\Conversation;

use BaseFrame\Exception\Domain\ParseFatalException;
use CompassApp\Domain\Member\Entity\Member;
use CompassApp\Domain\User\Main;
use CompassApp\Domain\Member\Struct\Main as MemberStruct;

/**
 * класс-интерфейс для таблицы company_data . member_list
 */
class Gateway_Db_CompanyData_MemberList extends Gateway_Db_CompanyData_Main {

	protected const _TABLE_KEY = "member_list";

	// -------------------------------------------------------
	// PUBLIC METHODS
	// -------------------------------------------------------

	/**
	 * метод для получения списка участников по маске группы
	 *
	 * @param int $permissions
	 *
	 * @return MemberStruct[]
	 * @throws ParseFatalException
	 */
	public static function getByPermissionMask(int $permissions):array {

		// запрос проверен на EXPLAIN (INDEX=npc_type.role.company_joined_at)
		$query = "SELECT * FROM `?p` FORCE INDEX (`?p`) WHERE `npc_type` = ?i AND `role` = ?i AND `permissions` & ?i LIMIT ?i";
		$rows  = ShardingGateway::database(self::_DB_KEY)->getAll(
			$query, self::_TABLE_KEY, "npc_type.role.company_joined_at", Main::NPC_TYPE_HUMAN, Member::ROLE_ADMINISTRATOR, $permissions, 10000);

		$list = [];
		foreach ($rows as $row) {
			$list[$row["user_id"]] = self::_formatRow($row);
		}

		return $list;
	}
	// -------------------------------------------------------
	// PROTECTED
	// -------------------------------------------------------

	/**
	 * преобразовываем массив в структуру
	 *
	 */
	protected static function _formatRow(array $row):MemberStruct {

		return new MemberStruct(
			$row["user_id"],
			$row["role"],
			$row["npc_type"],
			$row["permissions"],
			$row["created_at"],
			$row["updated_at"],
			$row["company_joined_at"],
			$row["left_at"],
			$row["full_name_updated_at"],
			$row["full_name"],
			$row["mbti_type"],
			$row["short_description"],
			$row["avatar_file_key"],
			$row["comment"],
			fromJson($row["extra"]),
		);
	}
}